### PR TITLE
Develop upstream sync 190603

### PR DESCRIPTION
**Pull request by parallelo**
Merge log:
```
$ git merge upstream/master --no-edit
Auto-merging tensorflow/workspace.bzl
Auto-merging tensorflow/tools/api/golden/v1/tensorflow.test.pbtxt
Auto-merging tensorflow/tensorflow.bzl
Auto-merging tensorflow/python/ops/nn_test.py
Auto-merging tensorflow/python/kernel_tests/relu_op_test.py
Auto-merging tensorflow/python/kernel_tests/control_flow_ops_py_test.py
Auto-merging tensorflow/python/kernel_tests/BUILD
Auto-merging tensorflow/python/keras/distribute/BUILD
Auto-merging tensorflow/python/keras/backend_test.py
Auto-merging tensorflow/python/keras/BUILD
Auto-merging tensorflow/python/framework/test_util.py
Auto-merging tensorflow/python/eager/backprop_test.py
Auto-merging tensorflow/python/eager/BUILD
Auto-merging tensorflow/python/distribute/BUILD
Auto-merging tensorflow/python/BUILD
Auto-merging tensorflow/lite/tutorials/post_training_integer_quant.ipynb
Auto-merging tensorflow/lite/python/lite_test.py
Removing tensorflow/core/profiler/README.md
Auto-merging tensorflow/core/ops/nn_ops.cc
CONFLICT (content): Merge conflict in tensorflow/core/ops/nn_ops.cc
Auto-merging tensorflow/core/kernels/strided_slice_op.cc
Auto-merging tensorflow/core/kernels/scatter_nd_op.cc
Auto-merging tensorflow/core/kernels/resource_variable_ops.cc
Auto-merging tensorflow/core/kernels/partitioned_function_ops.cc
Removing tensorflow/core/kernels/mkl_conv_grad_bias_ops.cc
Auto-merging tensorflow/core/kernels/fused_batch_norm_op.cc
CONFLICT (content): Merge conflict in tensorflow/core/kernels/fused_batch_norm_op.cc
Auto-merging tensorflow/core/kernels/data_format_ops.cc
Auto-merging tensorflow/core/kernels/cwise_op_gpu_random_grad.cu.cc
Auto-merging tensorflow/core/kernels/conv_ops_gpu.h
Auto-merging tensorflow/core/kernels/conv_ops_3d.cc
Auto-merging tensorflow/core/kernels/conv_ops.cc
CONFLICT (content): Merge conflict in tensorflow/core/kernels/conv_ops.cc
Auto-merging tensorflow/core/kernels/conv_grad_ops_3d.cc
Auto-merging tensorflow/core/kernels/conv_grad_input_ops.cc
Auto-merging tensorflow/core/kernels/conv_grad_filter_ops.cc
Auto-merging tensorflow/core/kernels/conv_2d_gpu.h
CONFLICT (content): Merge conflict in tensorflow/core/kernels/conv_2d_gpu.h
Auto-merging tensorflow/core/kernels/BUILD
Auto-merging tensorflow/core/grappler/optimizers/remapper.cc
Auto-merging tensorflow/core/grappler/optimizers/layout_optimizer.cc
Auto-merging tensorflow/core/BUILD
Auto-merging tensorflow/compiler/xla/tests/BUILD
Auto-merging tensorflow/compiler/xla/service/gpu/tests/BUILD
Auto-merging tensorflow/compiler/xla/service/gpu/stream_executor_util.h
Auto-merging tensorflow/compiler/xla/service/gpu/redzone_allocator.cc
Auto-merging tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc
Auto-merging tensorflow/compiler/xla/service/gpu/buffer_comparator.cc
CONFLICT (content): Merge conflict in tensorflow/compiler/xla/service/gpu/buffer_comparator.cc
Auto-merging tensorflow/compiler/xla/service/gpu/BUILD
CONFLICT (content): Merge conflict in tensorflow/compiler/xla/service/gpu/BUILD
Auto-merging tensorflow/compiler/tests/BUILD
Auto-merging configure.py
Auto-merging README.md
Automatic merge failed; fix conflicts and then commit the result.
```

### CODE REVIEW

File: tensorflow/core/kernels/conv_ops.cc [hunk 1]
```diff
@@ -563,21 +563,6 @@ template struct LaunchConv2DOp<CPUDevice, float>;
 template struct LaunchConv2DOp<CPUDevice, double>;
 
 #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
-
-#if GOOGLE_CUDA
-<<<<<<< HEAD
-// Returns true if the given StreamExecutor is for a Volta or newer nvidia GPU.
```
Comment (whchung) [inline tensorflow/core/kernels/conv_ops.cc]: @parallelo we do want to keep this function. please refer to #29185

Comment (parallelo) [reply tensorflow/core/kernels/conv_ops.cc]: Yes, agree.  But in the upstream repo, that function has been moved to `tensorflow/core/kernels/conv_ops_gpu.h`.  

Let me know if I'm still missing something.  

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream-sync-190603/tensorflow/core/kernels/conv_ops_gpu.h#L30-L37

Comment (whchung) [reply tensorflow/core/kernels/conv_ops.cc]: Thanks for clarification. Please disregard my earlier comment then.

Comment (parallelo) [reply tensorflow/core/kernels/conv_ops.cc]: Sure, no prob.  I appreciate your review of the changes.

**Comment by parallelo** [comment]
Looks like the stream_executor->Init() interface changed.  Will need to update the ROCm side in a similar way to the CUDA side:  
https://github.com/tensorflow/tensorflow/commit/47984a7d05937b978d7a80f2e9d60b2d32c06c07#diff-bfcdbead896609a7631e9a5bb0353393R176

**Comment by whchung** [comment]
please check if #29167 from @deven-amd should be cherry-picked.

**Comment by parallelo** [comment]
On my local machine, I'm just seeing a single failure so far.  Appears to be a non-registered Cholesky op.  
```
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot assign a device for operation while/switch_case/indexed_c
ase/branch0/_0/br0_Cholesky: Could not satisfy explicit device specification '/device:GPU:0' because no supported kernel for G
PU devices is available.
Colocation Debug Info:
Colocation group had the following types and supported devices: 
Root Member(assigned_device_name_index_=-1 requested_device_name_='/device:GPU:0' assigned_device_name_='' resource_device_nam
e_='' supported_device_types_=[CPU, XLA_CPU, XLA_GPU] possible_devices_=[]
Cholesky: CPU XLA_CPU XLA_GPU 
```
I'll try to add it tonight and kick off another test cycle.

**Comment by parallelo** [comment]
For `rocm-xla`, here was the result:
```
Executed 521 out of 522 tests: 520 tests pass, 1 fails to build
//tensorflow/compiler/tests:image_ops_test_gpu                           FAILED
```
And it appeared to have run out of memory:
```
  (0) Resource exhausted: Failed to allocate request for 1.46GiB (1573524800B) on device ordinal 0
         [[{{node ResizeBilinear}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.
```
So.. let's reboot that CI node and do another run of the xla test -- 

jenkins: retest rocm-xla please

**Comment by parallelo** [comment]
Updated stats after the latest test cycle --

### rocm
```
Executed 1145 out of 1145 tests: 1142 tests pass and 3 fail locally.
```
```
//tensorflow/python/keras/distribute:keras_embedding_model_correctness_test TIMEOUT in 5 out of 5 in 915.0s
//tensorflow/python:control_flow_ops_test                                FAILED in 5 out of 5 in 116.8s
//tensorflow/python:control_flow_ops_test_gpu                        FAILED in 5 out of 5 in 35.0s
```
### rocm-v2
```
Executed 1127 out of 1127 tests: 1120 tests pass and 7 fail locally.
```
```
//tensorflow/python/keras:integration_test                               FLAKY, failed in 4 out of 5 in 915.0s
//tensorflow/python/keras/distribute:keras_embedding_model_correctness_test FLAKY, failed in 4 out of 5 in 915.0s
//tensorflow/python:control_flow_ops_test                                FAILED in 5 out of 5 in 175.5s
//tensorflow/python:control_flow_ops_test_gpu                            FAILED in 5 out of 5 in 77.7s
//tensorflow/python/distribute:minimize_loss_test                        FAILED in 5 out of 5 in 203.9s
//tensorflow/python/distribute:minimize_loss_test_gpu                    FAILED in 5 out of 5 in 148.8s
//tensorflow/python/distribute:step_fn_test                              FAILED in 5 out of 5 in 47.7s
//tensorflow/python/distribute:step_fn_test_gpu                          FAILED in 5 out of 5 in 65.6s
//tensorflow/python/keras/distribute:keras_utils_test                    FAILED in 5 out of 5 in 576.0s
```
### rocm-xla
```
Executed 521 out of 522 tests: 520 tests pass, 1 fails to build
//tensorflow/compiler/tests:image_ops_test_gpu                           FAILED
```
And it appeared to have run out of memory:
```
  (0) Resource exhausted: Failed to allocate request for 1.46GiB (1573524800B) on device ordinal 0
         [[{{node ResizeBilinear}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.
```
### cuda
```
Executed 1202 out of 1202 tests: 1202 tests pass.
```

**Comment by parallelo** [comment]
### Triaging / troubleshooting `rocm` and `rocm-v2` failing unit tests

#### `//tensorflow/python/keras:integration_test`
Flaky/timeout.  Change `size = large`

#### `//tensorflow/python/keras/distribute:keras_embedding_model_correctness_test`
Flaky/timeout.  Change `size = large`

#### `//tensorflow/python:control_flow_ops_test*`
Add `no_rocm` tag due to dependency on unhandled Cholesky op

#### `//tensorflow/python/distribute:*`
Still investigating this error signature: 
```
ERROR: testTrainNetwork_test_distribution_MirroredCPUAndGPU_optimizerfn_GradientDescentV1_mode_graph_usecallableloss_False (__main__.MinimizeLossStepTest)
testTrainNetwork_test_distribution_MirroredCPUAndGPU_optimizerfn_GradientDescentV1_mode_graph_usecallableloss_False (__main__.MinimizeLossStepTest)
testTrainNetwork_test_distribution_MirroredCPUAndGPU_optimizerfn_GradientDescentV1_mode_graph_usecallableloss_False(distribution=MirroredCPUAndGPU, use_callable_loss=False, optimizer_fn=GradientDescentV1, mode='graph')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_prj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/distribute/minimize_loss_test_gpu.runfiles/absl_py/absl/testing/parameterized.py", line 262, in bound_param_test
    test_method(self, **testcase_params)
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_prj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/distribute/minimize_loss_test_gpu.runfiles/org_tensorflow/tensorflow/python/distribute/combinations.py", line 183, in decorated
    test_method(**kwargs_to_pass)
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_prj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/distribute/minimize_loss_test_gpu.runfiles/org_tensorflow/tensorflow/python/distribute/minimize_loss_test.py", line 104, in testTrainNetwork
    run_step = sess.make_callable(run_step())
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_prj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/distribute/minimize_loss_test_gpu.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1234, in make_callable
    self._extend_graph()
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_prj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/distribute/minimize_loss_test_gpu.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1374, in _extend_graph
    tf_session.ExtendSession(self._session)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot place the graph because a reference or resource edge connects colocation groups with incompatible assigned devices: /job:localhost/replica:0/task:0/device:CPU:0 vs /job:localhost/replica:0/task:0/device:GPU:0. The edge src node is while/exit/_35 , and the dst node is while
```
#### `//tensorflow/python/keras/distribute:keras_utils_test`
Still investigating this error signature: 
```
ERROR: test_batchnorm_correctness_test_distribution_Default_mode_graph_fused_True_cloning_
False_optimizer_GradientDescentKerasV2 (__main__.TestDistributionStrategyWithNormalization
Layer)
test_batchnorm_correctness_test_distribution_Default_mode_graph_fused_True_cloning_False_o
ptimizer_GradientDescentKerasV2 (__main__.TestDistributionStrategyWithNormalizationLayer)
test_batchnorm_correctness_test_distribution_Default_mode_graph_fused_True_cloning_False_o
ptimizer_GradientDescentKerasV2(distribution=Default, optimizer=GradientDescentKerasV2, mo
de='graph', fused=True, cloning=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_p
rj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin
/tensorflow/python/keras/distribute/keras_utils_test.runfiles/org_tensorflow/tensorflow/py
thon/client/session.py", line 1356, in _do_call
    return fn(*args)
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_p
rj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin
/tensorflow/python/keras/distribute/keras_utils_test.runfiles/org_tensorflow/tensorflow/py
thon/client/session.py", line 1339, in _run_fn
    self._extend_graph()
  File "/home/prj47-rack-51/tensorflow-upstream/bazel-ci_build-cache/.cache/bazel/_bazel_p
rj47-rack-51/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/bin
/tensorflow/python/keras/distribute/keras_utils_test.runfiles/org_tensorflow/tensorflow/py
thon/client/session.py", line 1374, in _extend_graph
    tf_session.ExtendSession(self._session)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Node 'SGD/gradients/batch_no
rmalization/cond_grad/If': Connecting to invalid output 4 of source node batch_normalizati
on/cond which has 3 outputs
```

**Comment by parallelo** [comment]
Note that all of `rocm-v2`'s `distribute` unit tests are giving the following warning about "non-GPU devices" and avoiding NCCL all-reduce:  
```
W0604 18:30:22.603292 140652288788224 cross_device_ops.py:1168] There is non-GPU devices in `tf.distribute.Strategy`, not using nccl allreduce.
```
In terms of the upstream merge process, I think it makes sense to disable these until we can take a deeper look at what's happening across the distribute tests.

**Comment by parallelo** [comment]
At this point, we appear to have a single unit test failure left.  It belongs to `rocm-xla`:  
```
In file included from tensorflow/compiler/jit/tests/auto_clustering_test.cc:17:0:
./tensorflow/compiler/jit/tests/auto_clustering_test_helper.h:56:36: error: 'benchmark' has not been declared
                                    benchmark::State& state);
 ```

The `auto_clustering_test` appears to have been added to upstream within the last week [here](https://github.com/tensorflow/tensorflow/commit/cd4872bd01eb5ded253ccbefe434d20f4462b4cb).

**Comment by whchung** [comment]
@parallelo let's check if the broken build has been fixed upstream, or else we can come up with a fix?

it seems to have something to do with `tensorflow/core/platform/test_benchmark.h`.

**Comment by parallelo** [comment]
This recent commit likely has the fix:  https://github.com/tensorflow/tensorflow/commit/194f0f21bcc3fa835eb9dd6f79361d2f907a7f58#diff-936451e1615925063fe1d0b72f294a43

Testing now.

**Comment by parallelo** [comment]
All set now:
```
//tensorflow/compiler/jit/tests:auto_clustering_test                     PASSED in 6.9s
```
```
Executed 522 out of 522 tests: 522 tests pass.
```

**Comment by whchung** [comment]
Merging based on the fact `rocm-xla` is a success. Other paths were already passing in the previous build.

**Pull request merged**